xds/internal/xdsclient/authority.go

```diff
@@ -306,365 +334,281 @@ func (a *authority) updateResourceStateAndScheduleCallbacks(rType xdsresource.Ty
+				if a.logger.V(2) {
+					a.logger.Infof("%d more watchers exist for type %q, resource name %q", rType.TypeName(), resourceName)
+				}
+				return
+			}
 
-func (a *authority) triggerResourceNotFoundForTesting(rType xdsresource.Type, resourceName string) {
-	a.resourcesMu.Lock()
-	defer a.resourcesMu.Unlock()
+			// There are no more watchers for this resource. Unsubscribe this
+			// resource from all channels where it was subscribed to and delete
+			// the state associated with it.
+			if a.logger.V(2) {
+				a.logger.Infof("Removing last watch for resource name %q", resourceName)
+			}
+			for _, xc := range state.xdsChannelConfigs {
+				xc.xc.unsubscribe(rType, resourceName)
+			}
+			delete(resources, resourceName)
 
-	if a.closed {
-		return
-	}
-	resourceStates := a.resources[rType]
-	state, ok := resourceStates[resourceName]
-	if !ok {
-		return
-	}
-	// if watchStateTimeout already triggered resource not found above from
-	// normal watch expiry.
-	if state.wState == watchStateCanceled || state.wState == watchStateTimeout {
-		return
-	}
-	state.wState = watchStateTimeout
-	state.cache = nil
-	state.md = xdsresource.UpdateMetadata{Status: xdsresource.ServiceStatusNotExist}
-	for watcher := range state.watchers {
-		watcher := watcher
-		a.serializer.TrySchedule(func(context.Context) { watcher.OnResourceDoesNotExist(func() {}) })
-	}
+			// If there are no more watchers for this resource type, delete the
+			// resource type from the top-level map.
+			if len(resources) == 0 {
+				if a.logger.V(2) {
+					a.logger.Infof("Removing last watch for resource type %q", rType.TypeName())
+				}
+				delete(a.resources, rType)
+			}
+			// If there are no more watchers for any resource type, release the
+			// reference to the xdsChannels.
+			if len(a.resources) == 0 {
+				if a.logger.V(2) {
+					a.logger.Infof("Removing last watch for for any resource type, releasing reference to the xdsChannel")
+				}
+				a.closeXDSChannelsLocked()
+			}
+		}, func() { close(done) })
+		<-done
+	})
 }
 
-// sendDiscoveryRequestLocked sends a discovery request for the specified
-// resource type and resource names. Even though this method does not directly
-// access the resource cache, it is important that `resourcesMu` be held when
-// calling this method to ensure that a consistent snapshot of resource names is
-// being requested.
-func (a *authority) sendDiscoveryRequestLocked(rType xdsresource.Type, resources map[string]*resourceState) {
-	resourcesToRequest := make([]string, len(resources))
-	i := 0
-	for name := range resources {
-		resourcesToRequest[i] = name
-		i++
+// xdsChannelToUseLocked returns the xdsChannel to use for communicating with
+// the management server. If an active channel is available, it returns that.
+// Otherwise, it creates a new channel using the first server configuration in
+// the list of configurations, and returns that.
+func (a *authority) xdsChannelToUseLocked() *xdsChannelWithConfig {
```

Comment:
For the onLocked method here and below; mention what mu needs to be held while calling these.

Reply:
Okay. Should we drop the "Locked" suffix then?